File: Data Structures & Algorithms/count-number-of-islands/submission-1.cpp
class Solution {
public:
void bfs(vector<vector<char>>& grid, int r, int c, int rows, int cols){
        queue <pair<int,int>> q;
        q.push({r,c});
        grid[r][c] = '1';
        while(!q.empty()){
            auto [row, col] = q.front();
            q.pop();

            if(row+1<rows && grid[row+1][col] == '1'){
                grid[row+1][col] = '#';
                q.push({row+1, col});
            }
            if(row-1>=0 && grid[row-1][col] == '1'){
                grid[row-1][col] = '#';
                q.push({row-1, col});
            }
            if(col+1<cols && grid[row][col+1] == '1'){
                grid[row][col+1] = '#';
                q.push({row, col+1});
            }
            if(col-1>=0 && grid[row][col-1] == '1'){
                grid[row][col-1] = '#';
                q.push({row, col-1});
            }
        }
    }
    
    int numIslands(vector<vector<char>>& grid) {
        int rows = grid.size(), cols = grid[0].size(), count =0;
        for(int i=0; i<rows; i++){
            for(int j=0; j<cols; j++){
                if(grid[i][j]=='1'){
                count++;
                bfs(grid, i,j, rows, cols);
                }
            }
        }
        return count;

    }
};

File: Data Structures & Algorithms/rotting-fruit/submission-1.py
from collections import deque
class Solution:
    def orangesRotting(self, grid: List[List[int]]) -> int:
        queue = deque()
        m = len(grid)
        n = len(grid[0])

        nr = [0,-1,1,0]
        nc = [-1,0,0,1]

        fresh = 0
        for i in range(0,m):
            for j in range(0,n):
                if grid[i][j] == 2:
                    queue.append((i,j))
                elif grid[i][j] == 1:
                    fresh += 1
        count = 0
        while queue and fresh:
            for _ in range(len(queue)):
                r, c = queue.popleft()
                for i in range(0,4):
                    dr = r + nr[i]
                    dc = c + nc[i]
                    if dr >= 0 and dr < m and dc >= 0 and dc < n and grid[dr][dc] == 1:
                        grid[dr][dc] = grid[r][c] + 1
                        queue.append((dr,dc))
                        fresh -= 1
            count = count + 1
        return count if fresh == 0 else -1